ProcessLib/DensityDrivenFlow/DensityDrivenFlowFEM.h

```diff
@@ -0,0 +1,250 @@
+            double thermal_conductivity =
+                thermal_conductivity_solid * (1 - porosity) +
+                thermal_conductivity_fluid * porosity;
+
+            auto const num_nodes = ShapeFunction::NPOINTS;
+            auto const& sm = _shape_matrices[ip];
+            auto const& wp = _integration_method.getWeightedPoint(ip);
+            auto Ktt = local_K.template block<num_nodes, num_nodes>(0, 0);
+            auto Mtt = local_M.template block<num_nodes, num_nodes>(0, 0);
+            auto Kpp = local_K.template block<num_nodes, num_nodes>(num_nodes,
+                                                                    num_nodes);
+            auto Mpp = local_M.template block<num_nodes, num_nodes>(num_nodes,
+                                                                    num_nodes);
+            auto Bp = local_b.template block<num_nodes, 1>(num_nodes, 0);
+
+            double T_int_pt = 0.0;
+            double p_int_pt = 0.0;
+            // Order matters: First T, then P!
+            NumLib::shapeFunctionInterpolate(local_x, sm.N, T_int_pt, p_int_pt);
+
+            double const delta_T(T_int_pt - temperature0);
+            // TODO include this via material lib
+            double density_water_T = density_fluid * (1 - beta * delta_T);
+            // TODO include viscosity computations via material lib
+            // double const viscosity = viscosity0 * std::exp(- delta_T/75.0);
+            double hydraulic_conductivity = intrinsic_permeability / viscosity0;
+
+            auto p_nodal_values = Eigen::Map<const Eigen::VectorXd>(
+                &local_x[num_nodes], num_nodes);
+
+            auto velocity =
+                (-hydraulic_conductivity * sm.dNdx * p_nodal_values).eval();
+
+            if (_process_data.has_gravity)
+                velocity += hydraulic_conductivity * density_water_T * b;
+
+            auto const integral_term = sm.detJ * wp.getWeight();
+            // matrix assembly
+            Ktt.noalias() +=
+                integral_term *
+                (sm.dNdx.transpose() * thermal_conductivity * sm.dNdx +
+                 sm.N.transpose() * velocity.transpose() * sm.dNdx *
+                     density_fluid * specific_heat_capacity_fluid);
+            Kpp.noalias() +=
+                integral_term * sm.dNdx.transpose() *
+                /* Real_hydraulic_conductivity */ hydraulic_conductivity *
+                sm.dNdx;
+            Mtt.noalias() +=
+                integral_term * sm.N.transpose() * heat_capacity * sm.N;
+            Mpp.noalias() += integral_term * sm.N.transpose() * storage * sm.N;
+            if (_process_data.has_gravity)
+                Bp += hydraulic_conductivity * integral_term *
+                      sm.dNdx.transpose() * b * density_water_T;
+            /* with Oberbeck-Boussing assumption density difference only exists
+             * in buoyancy effects */
+
+            // velocity computed for output.
```

Comment:
The velocity for output have been calculated again in computeSecondaryVariableConcrete. Otherwise, the output velocity is from the previous nonlinear iteration or from the previous time step. Therefore following loop has be move to computeSecondaryVariableConcrete too.